### PR TITLE
Fixes #837

### LINKED ISSUE #837: `Graph.subjects` and `Graph.predicates` should return only unique values
When calling the `Graph` method `subjects` or `predicates`, the generator should only return unique values.  For example:

```python
from rdflib import Graph
graph = Graph()
graph.parse("https://www.w3.org/People/Berners-Lee/card")
list(graph.subjects())
```

returns the following:
```
[rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card'),
 rdflib.term.BNode('N170b6415b06f45c98cb7a09398bbe818'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('http://www.w3.org/data#W3C'),
 rdflib.term.URIRef('http://www.ecs.soton.ac.uk/~dt2/dlstuff/www2006_data#panel-panelk01'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('http://wiki.ontoworld.org/index.php/_IRW2006'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('http://www.w3.org/2011/Talks/0331-hyderabad-tbl/data#talk'),
 rdflib.term.BNode('Nb0ee70dcdcc042609d1dde9df6b9efd5'),
 rdflib.term.BNode('Nebae90d841e441adbf72a3836347c354'),
 rdflib.term.URIRef('https://timbl.com/timbl/Public/friends.ttl'),
 rdflib.term.URIRef('http://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('http://dig.csail.mit.edu/breadcrumbs/blog/4'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card'),
 rdflib.term.BNode('N112689303c124871942e42a7ded52e64'),
 rdflib.term.BNode('N112689303c124871942e42a7ded52e64'),
 rdflib.term.URIRef('http://dig.csail.mit.edu/data#DIG'),
 rdflib.term.URIRef('https://www.w3.org/DesignIssues/Overview.html'),
 rdflib.term.URIRef('http://dig.csail.mit.edu/2007/01/camp/data#course'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://timbl.com/timbl/Public/friends.ttl'),
 rdflib.term.BNode('Nebae90d841e441adbf72a3836347c354'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://timbl.com/timbl/Public/friends.ttl'),
 rdflib.term.BNode('Nce03d86469d64deb945313ef31a8e661'),
 rdflib.term.BNode('Nce03d86469d64deb945313ef31a8e661'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.BNode('N1d372c7ed22e4de5a167177739cc29a0'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/DesignIssues/Overview.html'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.BNode('N112689303c124871942e42a7ded52e64'),
 rdflib.term.URIRef('https://timbl.com/timbl/Public/friends.ttl'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('http://www.w3.org/2011/Talks/0331-hyderabad-tbl/data#talk'),
 rdflib.term.URIRef('http://www.w3.org/2000/10/swap/data#Cwm'),
 rdflib.term.URIRef('http://dig.csail.mit.edu/breadcrumbs/blog/4'),
 rdflib.term.URIRef('http://wiki.ontoworld.org/index.php/_IRW2006'),
 rdflib.term.URIRef('https://timbl.com/timbl/Public/friends.ttl'),
 rdflib.term.URIRef('http://www.ecs.soton.ac.uk/~dt2/dlstuff/www2006_data#panel-panelk01'),
 rdflib.term.BNode('N170b6415b06f45c98cb7a09398bbe818'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.BNode('Nce03d86469d64deb945313ef31a8e661'),
 rdflib.term.BNode('Nb0ee70dcdcc042609d1dde9df6b9efd5'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.BNode('Nebae90d841e441adbf72a3836347c354'),
 rdflib.term.BNode('N713bca5215e14aa4ae1a3dfca525ce37'),
 rdflib.term.BNode('N1d372c7ed22e4de5a167177739cc29a0'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.BNode('N713bca5215e14aa4ae1a3dfca525ce37'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.BNode('N112689303c124871942e42a7ded52e64'),
 rdflib.term.URIRef('http://dig.csail.mit.edu/breadcrumbs/blog/4'),
 rdflib.term.URIRef('http://dig.csail.mit.edu/2005/ajar/ajaw/data#Tabulator'),
 rdflib.term.BNode('N112689303c124871942e42a7ded52e64'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.BNode('N1d372c7ed22e4de5a167177739cc29a0'),
 rdflib.term.BNode('N1d372c7ed22e4de5a167177739cc29a0'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.BNode('N1d372c7ed22e4de5a167177739cc29a0')]
```

I would expect it to return something more along the lines of:

```
[rdflib.term.URIRef('http://www.ecs.soton.ac.uk/~dt2/dlstuff/www2006_data#panel-panelk01'),
 rdflib.term.URIRef('http://dig.csail.mit.edu/data#DIG'),
 rdflib.term.URIRef('http://www.w3.org/2000/10/swap/data#Cwm'),
 rdflib.term.URIRef('http://dig.csail.mit.edu/2005/ajar/ajaw/data#Tabulator'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card'),
 rdflib.term.URIRef('http://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.BNode('N170b6415b06f45c98cb7a09398bbe818'),
 rdflib.term.BNode('N1d372c7ed22e4de5a167177739cc29a0'),
 rdflib.term.URIRef('http://wiki.ontoworld.org/index.php/_IRW2006'),
 rdflib.term.BNode('N112689303c124871942e42a7ded52e64'),
 rdflib.term.URIRef('https://www.w3.org/DesignIssues/Overview.html'),
 rdflib.term.BNode('N713bca5215e14aa4ae1a3dfca525ce37'),
 rdflib.term.URIRef('http://www.w3.org/data#W3C'),
 rdflib.term.BNode('Nce03d86469d64deb945313ef31a8e661'),
 rdflib.term.URIRef('http://dig.csail.mit.edu/2007/01/camp/data#course'),
 rdflib.term.BNode('Nebae90d841e441adbf72a3836347c354'),
 rdflib.term.URIRef('http://www.w3.org/2011/Talks/0331-hyderabad-tbl/data#talk'),
 rdflib.term.BNode('Nb0ee70dcdcc042609d1dde9df6b9efd5'),
 rdflib.term.URIRef('https://www.w3.org/People/Berners-Lee/card#i'),
 rdflib.term.URIRef('https://timbl.com/timbl/Public/friends.ttl'),
 rdflib.term.URIRef('http://dig.csail.mit.edu/breadcrumbs/blog/4')]
```

While you could technically get the list of unique subjects using `list(set(graph.subjects()))` it would be nice if `subjects` did this automatically.

### PR DESCRIPTION
**Pull request by nilay77**
- Fixed Graph.subjects, Graph.predicates, and Graph.objects to return only unique values. 
- Added test case for the above 3 methods

### CODE REVIEW

File: rdflib/graph.py [hunk 1]
```diff
@@ -617,18 +617,27 @@ def set(self, triple):
 
     def subjects(self, predicate=None, object=None):
         """A generator of subjects with the given predicate and object"""
+        subs = []
```
Comment (niklasl) [inline rdflib/graph.py:620]: Use sets for keeping track of seen (lists have O(n) lookup time).
```suggestion
        subs = set()
```

(And subsequently use `subs.add(s)`).

Comment (nilay77) [reply rdflib/graph.py:620]: Oh yes, this makes sense. I'll update this. Are any other changes required?

Comment (nilay77) [reply rdflib/graph.py]: @niklasl I have pushed the suggested changes.

**Comment by tgbugs** [comment]
This cannot go into core as is because the memory usage will explode for large graphs. As mentioned in the discussion of #837 this is something that probably has to be implemented on the stores or under a different name, or with a different set of arguments e.g. `def subjects(self, predicate=None, object=None, unique=False)` or `def subjects_unique(self, predicate=None, object=None)`.

**Comment by nilay77** [comment]
@tgbugs I have added a parameter named unique to the current methods as suggested by you. I have also added exception handling code which suggests setting the parameter unique to False. If you could give me an idea about how to produce a memory error I can even add a test case.

File: rdflib/graph.py [hunk 2]
```diff
@@ -615,20 +615,56 @@ def set(self, triple):
...
-            yield p
-
-    def objects(self, subject=None, predicate=None):
-        """A generator of objects with the given subject and predicate"""
-        for s, p, o in self.triples((subject, predicate, None)):
-            yield o
+    def subjects(self, predicate=None, object=None, unique=False):
+        """A generator of unique subjects with the given predicate and object"""
+
+        if(unique==False):
```
Comment (nicholascar) [inline rdflib/graph.py:621]: change to `if not unique:`

File: rdflib/graph.py [hunk 3]
```diff
@@ -615,20 +615,56 @@ def set(self, triple):
...
+                    try:
+                        subs.add(s)
+                    except MemoryError:
+                        print("MemoryError encountered. Consider setting parameter 'unique' to False")
+                        break
+
+    def predicates(self, subject=None, object=None, unique=False):
+        """A generator of unique predicates with the given subject and object"""
+
+        if(unique==False):
```
Comment (nicholascar) [inline rdflib/graph.py:638]: change to `if not unique:`

File: rdflib/graph.py [hunk 4]
```diff
@@ -615,20 +615,56 @@ def set(self, triple):
...
+                    try:
+                        preds.add(p)
+                    except MemoryError:
+                        print("MemoryError encountered. Consider setting parameter 'unique' to False")
+                        break
+
+    def objects(self, subject=None, predicate=None, unique=False):
+        """A generator of unique objects with the given subject and predicate"""
+
+        if(unique==False):
```
Comment (nicholascar) [inline rdflib/graph.py:655]: change to `if not unique:`

**Comment by nicholascar** [comment]
Closing this PR as the proponent hasn't addressed comments in a couple of months. @nilay77 please reopen if you want to progress this, but you will need to address comments.

**Pull request closed without merging**